model/apps/rtp-header.cc

```diff
@@ -58,140 +91,597 @@ TypeId RtpHeader::GetInstanceTypeId () const
+        const uint16_t endSeq = start.ReadNtohU16 ();
+        len_left -= 4;
+        const uint32_t nMetricBlocks = uint32_t (endSeq - beginSeq) + 1; //this wraps properly
+        NS_ASSERT (nMetricBlocks <= 0xffff);// length of 65536 not supported
+        const uint32_t nPaddingBlocks = nMetricBlocks % 2;
+        NS_ASSERT (len_left >= nMetricBlocks + nPaddingBlocks);
+        uint16_t seq = beginSeq;
+        for (auto i = 0; i < nMetricBlocks; ++i) {
+            const auto octet1 = start.ReadU8 ();
+            const auto octet2 = start.ReadU8 ();
+            if (RtpHdrGetBit (octet1, 7)) {
+                auto &mb = rb[seq];
+                mb.m_ecn = (octet1 >> 5) & 0x03;
+                uint16_t ato = (uint16_t (octet1) << 8) & 0x1f00;
+                ato |= uint16_t (octet2);
+                mb.m_ato = ato;
+            }
+            ++seq;
+        }
+        len_left -= nMetricBlocks;
+        if (nPaddingBlocks == 1) {
+            start.ReadNtohU16 (); //skip padding
+            --len_left;
+        }
+    }
+    // TODO (next patch): convert from NTP time to Us
+    m_latestTsUs = start.ReadNtohU32 ();
+    // Populate all timestamps once Report Timestamp is known
+    // TODO: report to authors
+    for (auto& rb : m_reportBlocks) {
+        for (auto& mb : rb.second) {
+            mb.second.m_timestampUs = AtoToTs (mb.second.m_ato);
+        }
+    }
     return GetSerializedSize ();
 }
 
-void FeedbackHeader::Print (std::ostream &os) const
+void CCFeedbackHeader::Print (std::ostream& os) const
+{
+    NS_ASSERT (m_length >= 2);
+    RtcpHeader::PrintN (os);
+    size_t i = 0;
+    for (const auto& rb : m_reportBlocks) {
+        const auto beginStop = CalculateBeginStopSeq (rb.second);
+        const uint16_t beginSeq = beginStop.first;
+        const uint16_t stopSeq = beginStop.second;
+        os << ", report block #" << i << " = "
+           << "{ SSRC = " << rb.first
+           << " [" << beginSeq << ".." << (stopSeq - 1) << "] --> ";
+        for (uint16_t j = beginSeq; j != stopSeq; ++j) {
+            const auto& mb_it = rb.second.find (j);
+            const bool received = (mb_it != rb.second.end ());
+            os << "<L=" << uint8_t (received);
+            if (received) {
+                const auto& mb = mb_it->second;
+                os << ", ECN=0x" << std::hex << mb.m_ecn
+                   << ", ATO=0x" << TsToAto (mb.m_timestampUs);
+            }
+            os << ">,";
+        }
+        os << " }, ";
+        ++i;
+    }
+    // TODO (next patch): convert from Us to NTP time
+    os << "report timestamp = " << m_latestTsUs << std::endl;
+}
+
+std::pair<uint16_t, uint16_t>
+CCFeedbackHeader::CalculateBeginStopSeq (const ReportBlock_t& rb)
+{
+    NS_ASSERT (!rb.empty ()); // at least one metric block
+    auto mb_it = rb.begin ();
+    const uint16_t first = mb_it->first;
+    if (rb.size () == 1) {
+        return std::make_pair (first, first + 1);
+    }
+    //calculate biggest gap
```

Comment:
what is the purpose of getting the biggest gap, only sample for the biggest gap?

Reply:
As the draft is currently worded, it is "legal" to include feedback for sequence ids scattered all over the uint16_t spectrum. So I need to detect the "biggest gap" considering sequence id wrapping. Once the "biggest gap" is detected, the lower bound of the gap will be the end sequence, and the higher bound will be the begin sequence of the report.
As agreed in our last meeting, I will write the authors suggesting that only sequence ids spanning one quadrant (16384 max sequence span) should be included in a single report. Which is quite reasonable, I think